### PR TITLE
q40603-proposal-attempt1

### PR DESCRIPTION
**Pull request by q40603**
---
A good PR should include the following parts:

- [ ] A clear title
- [ ] A meaningful message for PR, as well as its commits
- [ ] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [ ] excluding any irrelevant files, such as binaries, text files, or dot files
- [ ] passing tests/CI

### CODE REVIEW

**Comment by yungyuc** [comment]
Thanks @q40603 .  It seemed that you created the PR from `master` so that it contains unwanted changeset.  You should have created the PR from `proposal`.  To help you remove the unwanted change, I merged the `proposal` branch with `master`.  Could you do either of the following before I start review your proposal?

1. Pull and rebased to the current `proposal` branch.  Update the PR.
2. Use the new `master` / `proposal`, and create another PR.

**Comment by q40603** [comment]
Fix it with rebase.

File: proposal/q40603/proposal.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+Proposal for Pairs Trading 
+===========================
+
+A pairs trade is a trading strategy that involves matching a long position with a short position in two stocks with a high correlation. It uses statistical and technical analysis to seek out potential market-neutral profits.
+
+In this term project, I would like to optmize the Pairs Trading computing engine developed by the classmate in my lab. He combined the pairs trading techinque with Taiwan Stock Market intraday high-frequency trading, whcih I think is an opportunity and is suitalbe for NSD term project.
+
+Basic information
+=================
+
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:11]: Please add the github link in the proposal.

File: proposal/q40603/proposal.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
+Proposal for Pairs Trading
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: The template file uses [reStructuredText](https://docutils.sourceforge.io/rst.html), not markdown.  Is the filename a typo?

If you choose markdown, please reformat the file.

File: proposal/q40603/proposal.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+In this term project, I would like to optmize the Pairs Trading computing engine developed by the classmate in my lab. He combined the pairs trading techinque with Taiwan Stock Market intraday high-frequency trading, whcih I think is an opportunity and is suitalbe for NSD term project.
+
+Basic information
+=================
+
+Pairs Trading always comes along with a term "Cointegration". The basic ideas of Cointegration is to test whether two time series move similarly. When applied to stock market, we can find two assets that move with cointegration, assign different weight at the assets respectively, and sell the high priced stock, buy the low priced stock. 
+
+| two stock series | Coinegration series |
+| -------- | -------- |
+| ![](https://i.imgur.com/ajKsdUc.png)     | ![](https://i.imgur.com/2cPbJ7F.png)    |
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:16]: If you own the images, it's a good idea to add them here in the proposal directory as records.

File: proposal/q40603/proposal.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+
+
+
+
+
+Problem to solve
+================
+
+### Prerequiement
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Now this is markdown.  I am not pretty sure which text format you are trying to use.  It'd be better if you stick with one.

File: proposal/q40603/proposal.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+* Some of the stock suffers from bad liquidity, which means there exists empty value in our data. So we compute the average price within each minute, and backfill the data with empty value.
+* We collect the data between 9:16 ~ 11:46 to do cointegration test
+* Trade the stock from 11:46~13:30
+
+The original version of the Pairs Trading computing engine is written in Python, and it takes about 2 minutes to do the computing:
+* We have to test every two stock of top 150 company (11175 combinations)
+
+### Math Part
+
+待補
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Use English.

File: proposal/q40603/proposal.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+
+Perspective users
+=================
+
+Trader or the students under department of Finance.
+
+System architecture
+===================
+(TBD)
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Need to add architecture.

File: proposal/q40603/proposal.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+operations.
+
+Describe the system's work flow.  You may consider to use a flow chart but it
+is not required.
+
+Specify the constraints assume in your system.  Describe how it is modularized.
+
+API description
+===============
+(TBD)
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Need to add API.

File: proposal/q40603/proposal.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+Show how a user programmatically uses your system.  You are supposed to
+implement the system using both C++ and Python.  Describe how a user writes a
+script in the system.  If you provide API in both C++ and Python, describe
+both.
+
+Engineering infrastructure
+==========================
+(TBD)
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Add engineering.

File: proposal/q40603/proposal.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+Describe how you plan to put together the build system, testing framework, and
+documentation.  Show how you will do version control.
+
+You may use continuous integration, but it is not required.  If you use it,
+describe how it works in your code development.
+
+Schedule
+========
+(TBD)
```
Comment (yungyuc) [inline proposal/q40603/proposal.md]: Add schedule.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please correct the format and add the missing parts before the proposal due.

**Comment by yungyuc** [comment]
@q40603 When you finish updating your branch please add comments to let me know I should review again.

**Comment by q40603** [comment]
> @q40603 When you finish updating your branch please add comments to let me know I should review again.
Proposal finished.

File: proposal/q40603/proposal.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+* We collect the data between 9:16 ~ 11:46 to do cointegration test
+* Trade the stock from 11:46~13:30
+
+The original version of the Pairs Trading computing engine is written in Python, and it takes about 2 minutes to do the computing:
+* We have to test every two stock of top 150 company (11175 combinations)
+
+#### Math Part
+Consider a K-dimensional  Gaussian VAR(p) model with a trend component:
+
+![](https://i.imgur.com/gfkwosJ.png)
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:46]: The equations will be good to be included in a written report or presentation slides.  If you have not done it, you are encouraged to do it.

File: proposal/q40603/proposal.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+| -------- | -------- | -------- |
+| 4/7 ~ 4/30   | Input data processing, modularize the computation (design part)    |
+| 5/1 ~ 5/15 | ADF test ,VAR model, order select |
+| 5/15 ~ presentation| VECM weight calcualtion | 
+
+
+
+### Question
+
+1. In Numerical Software Development, are there any special design pattern for us to follow?
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:147]: This is an excellent question.  It's no and yes.  No, the design patterns for numerical, scientific, and engineering codes are are special.  But yes, the codes do use a special selection and combination of the patterns.

There is not a good reference for the numerical design patterns, but I think the patterns used in game design have good resemblance.  You may take a look at https://gameprogrammingpatterns.com/contents.html .

File: proposal/q40603/proposal.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+2. What is the best way to moduralize the whole computaion
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:148]: Your implementation is the starting line of the journey to discover what will be the ways.

File: proposal/q40603/proposal.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+3. In numerical computaion, many function relys on the previous function output, can we use stub software testing techinque in Numerical Software Devlopment?
```
Comment (yungyuc) [inline proposal/q40603/proposal.md:149]: It sounds like a good idea.  I do not see showstopper of using that technique.  Keep in mind that it doesn't solve all problems in the coupling.

**Comment by yungyuc** [comment]
LGTM.  I will merge once I have all proposal graded.

**Review by yungyuc — APPROVED** [review]

**Pull request merged**